### PR TITLE
PFT: introduce controller and initialize it

### LINKED ISSUE #48000: Product Form Template: migrate template to the database on WC install/upgrade
~Follow-up of #47999~

The source of truth for the Product Form Template is stored in the database, but the original definition is contained in the template file. Therefore, a mechanism must be created to synchronize and consolidate the data from the file to the dB.

[PoC PR](https://github.com/woocommerce/woocommerce/pull/46348/files)

### LINKED ISSUE #48203: Product Form Template: add a simple single-product form template file
Let's create a simple Product Template now, mainly for development purposes. Later, it will be updated once we tackle https://github.com/woocommerce/woocommerce/issues/47999

```html
<!-- wp:woocommerce/product-section {"title":"<?php _e( 'Basic details', 'woocommerce' ); ?>"} -->
<div data-block-name="woocommerce/product-section" class="wp-block-woocommerce-product-section" data-title="<?php _e( 'Basic details', 'woocommerce' ); ?>">
    <div>
        <!-- wp:woocommerce/product-regular-price-field -->
        <div data-block-name="woocommerce/product-regular-price-field" class="wp-block-woocommerce-product-regular-price-field"></div>
        <!-- /wp:woocommerce/product-regular-price-field -->
        <!-- wp:woocommerce/product-checkbox-field {"label":"<?php _e( 'Translatable Label', 'woocommerce' ); ?>","property":"testproperty"} -->
        <div data-block-name="woocommerce/product-checkbox-field" class="wp-block-woocommerce-product-checkbox-field" data-label="<?php _e( 'Translatable Label', 'woocommerce' ); ?>"></div>
        <!-- /wp:woocommerce/product-checkbox-field -->
    </div>
</div>
<!-- /wp:woocommerce/product-section -->
```

_PFT: **P**roduct **F**orm **T**emplate system_

### PR DESCRIPTION
**Pull request by retrofox**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR implements the mechanism to migrate the representation of a Product Form Template file from its content to the dB.

* This process runs when the WooCommerce plugin is installed or updated.
* The PR introduces a super `simple` product template. It will be addressed in a follow-up 48000
* It shows the Product Form CPT in the UI only for testing/developing process

Closes #48000
Closes #48203

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Since the process triggers when the plugin installs/updates, I've created a plugin that registers a WP CLI command.

1. Install the plugin

[pft-dev-tools](https://github.com/woocommerce/pft-dev-tools/archive/refs/heads/trunk.zip)

2. Active the plugin
3. Enable the `product-editor-template-system` feature flag
  a. Install the WooCommerce Beta Tester plugin
  b. Go to the Settings -> WCA Test Helper -> Features
  c. Enable the `product-editor-template-system` feature flag

4. Go to the Product Forms page `wp-admin/edit.php?post_type=product_form`
5. Confirm the `Simple` form post isn't there
image
6. Run the following WP CLI command to migrate the template definition to the dB

```sh
wp trigger upgrader_process_complete
```
7. Hard refresh
8. Confirm the Simple form post is there

Screenshot 2024-06-06 at 1 05 27 PM

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Review by nathanss — COMMENTED** [review]
When installing the plugin, I got the following message:

image

Is that fine?

Also, when editing the 'Simple' product-form, it shows like this in the post editor:

image

Is that expected?

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductFormsController.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		// Check if the action is an `update` or `install` action for a plugin.
+		if (
+			'install' !== $hook_extra['action'] &&
+			'update' !== $hook_extra['action'] ||
+			'plugin' !== $hook_extra['type']
+		) {
+			return;
+		}
+
+		$updated_plugins = $hook_extra['plugins'];
```
Comment (nathanss) [inline plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductFormsController.php]: This line seems to be causing an error when installing your test plugin.

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductFormsController.php [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+			return;
+		}
+
+		/**
+		 * Allow extend the list of templates that should be auto-generated.
+		 *
+		 * @since 9.1.0
+		 * @param array $templates List of templates to auto-generate.
+		 */
+		$templates = apply_filters(
```
Comment (nathanss) [inline plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductFormsController.php:78]: Is that something we want to make public right now? I've had a similar dilemma in [this PR](https://github.com/woocommerce/woocommerce/pull/47762/files#diff-c7e5447e92a9e360c5c5374feb6520a2602996e1f7a470a3574ac0bf0071af3bR104), where I marked a hook as experimental. Not sure if we have a correct process for new hooks.

**Review by nathanss — CHANGES_REQUESTED** [review]
Overall it looks good and works as expected! Just found an issue while installing the test plugin and left a question.

**Comment by retrofox** [comment]
> When installing the plugin, I got the following message:

It's curious why the [upgrader_process_complete](https://developer.wordpress.org/reference/hooks/upgrader_process_complete/) hook passes ~`undefined`~ `null` as the second argument, even though it's just an empty array.

> Is that fine?

It doesn't look fine at all 😆  The issue is due to the previous step.
Could you provide more details about the steps you took?
Could you tell me if you triggered the action by running the cli command, or something else?

> Also, when editing the 'Simple' product-form, it shows like this in the post editor:

This is expected. The problem is that the template is composed of Product Block Editor blocks that aren't available in the editor context. We may change this in the future.

For now, the templates are not editable. You have access to the edition page just for testing/developing purposes.

You can switch to edition mode to see the markup there. That's important for now.

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductFormsController.php (hunk 2 above)
Comment (retrofox) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductFormsController.php:78]: It's okay, considering the whole feature is hidden behind a feature flag. It's required for extensibility purposes.

**Review by nathanss — APPROVED** [review]
The latest changes fixed the issue while installing the plugin

**Pull request merged**